apps/src/dance/Dance.js

```diff
@@ -451,24 +455,41 @@ Dance.prototype.shouldShowSharing = function () {
  * This is called while this.p5 is in the preload phase.
  */
 Dance.prototype.onP5Preload = function () {
-  const getSelectedSong = () => getStore().getState().selectedSong;
-
   this.nativeAPI = new DanceParty(this.p5, {
-    getSelectedSong,
     onPuzzleComplete: this.onPuzzleComplete.bind(this),
     playSound: audioCommands.playSound,
     recordReplayLog: this.shouldShowSharing(),
   });
+  this.updateSongMetadata(getStore().getState().selectedSong);
   const spriteConfig = new Function('World', this.level.customHelperLibrary);
   this.nativeAPI.init(spriteConfig);
   this.nativeAPI.preload();
 };
 
+Dance.prototype.updateSongMetadata = function (id) {
+  this.songMetadataPromise = this.loadSongMetadata(id);
+};
+
+Dance.prototype.playSong = function (file, callback) {
+  audioCommands.playSound({url: file, callback: callback});
+};
+
+Dance.prototype.loadSongMetadata = async function (id) {
+  let songDataPath = '/api/v1/sound-library/hoc_song_meta';
+  const response = await fetch(`${songDataPath}/${id}.json`);
+  return await response.json();
+};
+
 /**
  * This is called while this.p5 is in the setup phase.
  */
 Dance.prototype.onP5Setup = function () {
+  this.preloadComplete = true;
   this.nativeAPI.setup();
+  this.p5setupPromiseResolve();
+  if (this.share) {
+    setTimeout(this.studioApp_.runButtonClick, 0);
```

Comment:
Is this `setTimeout` necessary?

Reply:
probably not, I was imitating some gamelab code, but I can just do it directly.